samples/subsys/canbus/canopen/README.rst

```diff
@@ -0,0 +1,353 @@
+   print("Power-on counter: {}".format(reboots.raw))
+
+   network.disconnect()
+
+Running the above Python code should produce the following output:
+
+.. code-block:: console
+
+   Initial heartbeat time: 1000 ms
+   Power-on counter: 1
+   Updated heartbeat time: 5000 ms
+   heartbeat time after reset: 1000 ms
+   Power-on counter: 2
+   Updated heartbeat time: 2000 ms
+   heartbeat time after store and reset: 2000 ms
+   Power-on counter: 3
+   heartbeat time after restore and reset: 1000 ms
+   Power-on counter: 4
+
+Note that the power-on counter value may be different.
+
+PDO Mapping
+===========
+Transmit Process Data Object (PDO) mapping for data at a given index
+of the CANopen object dictionary (here index ``0x2102``, the button
+press counter) can be accomplished using the following Python code:
+
+.. code-block:: py
+
+   import canopen
+   import os
+
+   ZEPHYR_BASE = os.environ['ZEPHYR_BASE']
+   EDS = os.path.join(ZEPHYR_BASE, 'samples', 'subsys', 'canbus', 'canopen',
+                      'objdict', 'objdict.eds')
+   NODEID = 10
+
+   network = canopen.Network()
+
+   # 'can0' and 'socketcan' is for GNU/Linux, please see the python-can documentation for other platforms:
+   network.connect(channel='can0', bustype='socketcan')
+
+   node = network.add_node(NODEID, EDS)
+   button = node.sdo['Button press counter']
+
+   # Read current TPDO mapping
+   node.tpdo.read()
+
+   # Enter pre-operational state to map TPDO
+   node.nmt.state = 'PRE-OPERATIONAL'
+
+   # Map TPDO 1 to transmit the button press counter on changes
+   node.tpdo[1].clear()
+   node.tpdo[1].add_variable('Button press counter')
+   node.tpdo[1].trans_type = 254
+   node.tpdo[1].enabled = True
+
+   # Save TPDO mapping
+   node.tpdo.save()
+   node.nmt.state = 'OPERATIONAL'
+
+   # Reset button press counter
+   button.raw = 0
+
+   print("Press the button 10 times")
+   while True:
+       node.tpdo[1].wait_for_reception()
+       print("Button press counter: {}".format(node.tpdo['Button press counter'].phys))
+       if node.tpdo['Button press counter'].phys >= 10:
+           break
+
+   network.disconnect()
+
+Running the above Python code should produce the following output:
+
+.. code-block:: console
+
+   Press the button 10 times
+   Button press counter: 0
+   Button press counter: 1
+   Button press counter: 2
+   Button press counter: 3
+   Button press counter: 4
+   Button press counter: 5
+   Button press counter: 6
+   Button press counter: 7
+   Button press counter: 8
+   Button press counter: 9
+   Button press counter: 10
+
+Modifying the Object Dictionary
+*******************************
+The CANopen object dictionary used in this sample application can be
+found under :zephyr_file:`samples/subsys/canbus/canopen/objdict` in
+the Zephyr tree. The object dictionary can be modified using any
+object dictionary editor supporting CANopenNode object dictionary code
+generation.
+
+A popular choice is the EDS editor from the `libedssharp`_
```

Comment:
Would it be possible to generate the object dictionary at the build time? Or is the dependency on libedssharp not acceptable?

Reply:
It is definitely possible to generate the object dictionary implementation files at build time using libedssharp. However, libedssharp is a C#/.NET application/library and, as far as I know, not readily available in binary form (dpkg, rpms, ...).

I am open to adding a dependency on it (since we already use it), but I think it would be a bit much to impose having a C#/.NET development environment setup on all Zephyr users.

Reply:
>  am open to adding a dependency on it (since we already use it), but I think it would be a bit much to impose having a C#/.NET development environment setup on all Zephyr users.

That is probably to much :smile: